### PR TITLE
Support ipv6 parsing

### PR DESCRIPTION
**Pull request by oronsh**
https://github.com/denoland/deno/issues/5261

### CODE REVIEW

File: std/http/server.ts [hunk 1]
```diff
@@ -252,9 +252,12 @@ export type HTTPOptions = Omit<Deno.ListenOptions, "transport">;
  *       req.respond({ body });
  *     }
  */
+
 export function serve(addr: string | HTTPOptions): Server {
   if (typeof addr === "string") {
-    const [hostname, port] = addr.split(":");
+    const addrParts = addr.split(":");
+    const port = addrParts[addrParts.length-1];
+    const hostname = addr.slice(0, addr.length - port.length - 1);
```
Comment (kitsonk) [inline std/http/server.ts:260]:
```suggestion
    const addrParts = addr.split(":");
    const port = parseInt(addrParts.pop()!, 10);
    const hostname = addrParts.length ? splitAddr.join(":") : undefined;
```

Comment (oronsh) [reply std/http/server.ts:260]: Thanks I'll fix these

**Comment by piscisaureus** [comment]
I think this notation super confusing and possibly ambiguous.
What about wrapping the ipv6 address in square brackets and like how it's done in URLS. https://en.wikipedia.org/wiki/IPv6_address#Literal_IPv6_addresses_in_network_resource_identifiers

**Comment by oronsh** [comment]
@piscisaureus I agree, seems like this is also the convention for rust std https://doc.rust-lang.org/std/net/struct.SocketAddrV6.html

**Comment by guzzur** [comment]
> I think this notation super confusing and possibly ambiguous.
> What about wrapping the ipv6 address in square brackets, like how it's done in URLs. https://en.wikipedia.org/wiki/IPv6_address#Literal_IPv6_addresses_in_network_resource_identifiers

Do we need/want to support passing the address with **and** without square brackets? I.e. should we support `Server.serve("[::1]:8080")` **and** `Server.serve(2001:0db8:85a3:0000:0000:8a2e:0370:7334:8080")`?

**Comment by oronsh** [comment]
I see that the `deno` is already parsing `ipv6` square brackets notation
https://github.com/denoland/deno/blob/master/cli/resolve_addr.rs#L17

The tests are working with `[]` or without it, so I assume it's just a matter of user preference.
WDYT? @piscisaureus @kitsonk @guzzur

**Comment by Raqbit** [comment]
In Go it is also possible to listen on port-only "addresses" like `:8080`, which I believe will listen on both `0.0.0.0:8080` and `[::]:8080`. 
This last feature has to do with how `net.Listen("tcp", ...)` works though, as Go also allows creating a `tcp6` or `tcp4` listener.

**Comment by bartlomieju** [comment]
@nayeemrmn could you please review?

File: std/http/server.ts [hunk 2]
```diff
@@ -252,9 +252,12 @@ export type HTTPOptions = Omit<Deno.ListenOptions, "transport">;
...
  *     }
  */
+
 export function serve(addr: string | HTTPOptions): Server {
   if (typeof addr === "string") {
-    const [hostname, port] = addr.split(":");
+    const addrParts = addr.split(":");
+    const port = parseInt(addrParts.pop()!, 10);
+    const hostname = addrParts.length ? addrParts.join(":") : undefined;
     addr = { hostname, port: Number(port) };
```
Comment (nayeemrmn) [inline std/http/server.ts:261]: Can we use the `URL` constructor here? It has IPv6 support.

```suggestion
    let url;
    try {
      url = new URL(`http://${addr}`);
    } catch {
      throw new TypeError("Invalid address.")
    }
    if (url.pathname || url.search || url.hash) {
      throw new TypeError("Invalid address.");
    }
    addr = { hostname: url.hostname, port: Number(url.port) };
```
Also, this should be extracted and applied to `listenAndServe()` as well.

Comment (oronsh) [reply std/http/server.ts:261]: Will do, why did you add `http://` in `new URL`? shouldn't it be part of the addr string?

Comment (nayeemrmn) [reply std/http/server.ts:261]: > Will do, why did you add `http://` in `new URL`? shouldn't it be part of the addr string?

No? It wasn't before. Doesn't make sense for the protocol to be part of the address here.

Comment (oronsh) [reply std/http/server.ts:261]: So how does the user pass the protocol? in `HttpOptions`? the user has an option to pass a string to the fn.
Can you please explain? :)

Comment (nayeemrmn) [reply std/http/server.ts:261]: The protocol is `http`... why would they pass a possibly different one?

Comment (oronsh) [reply std/http/server.ts:261]: What about `https`?

Comment (oronsh) [reply std/http/server.ts:261]: Oh sorry, `https` is `serveTLS` right?

Comment (oronsh) [reply std/http/server.ts:261]: Another question @nayeemrmn , why should this logic be applied to `listenAndServe` as well? it always calls `serve` after all...

Comment (nayeemrmn) [reply std/http/server.ts:261]: > why should this logic be applied to `listenAndServe` as well? it always calls `serve` after all...

My mistake, you're right. No need to extract for `listenAndServe()`.

File: std/http/server.ts [hunk 3]
```diff
@@ -237,6 +237,30 @@ export class Server implements AsyncIterable<ServerRequest> {
...
 
+/**
+ * Parse addr from string
+ *
+ *     const addr = "::1:8000";
+ *     parseAddrFromString(addr);
+ *
+ * @param addr Address string
+ */
+export function parseAddrFromStr(addr: string): HTTPOptions {
```
Comment (nayeemrmn) [inline std/http/server.ts:248]: This should be private, rename it to `_parseAddrFromStr()` if it must be exported.

Comment (oronsh) [reply std/http/server.ts:248]: 👍

Comment (oronsh) [reply std/http/server.ts]: @nayeemrmn , pushed changes, how does it look to you?

Comment (oronsh) [reply std/http/server.ts]: If all is good, I'd love to squash commits merge this PR

File: std/http/server.ts [hunk 4]
```diff
@@ -247,10 +275,10 @@ export type HTTPOptions = Omit<Deno.ListenOptions, "transport">;
  *       req.respond({ body });
  *     }
  */
+
 export function serve(addr: string | HTTPOptions): Server {
```
Comment (ry) [inline std/http/server.ts:279]: Please don't add a newline here. I think it might break the JSDoc.

Comment (oronsh) [reply std/http/server.ts:278]: Fixed

**Comment by oronsh** [comment]
@ry @kitsonk @nayeemrmn This is ready for another review 🙏

**Pull request merged**